### DISCUSSION TITLE
Directly modifying problems

### DISCUSSION THREAD

**Discussion by haz** [aig-upf/tarski#133] [Q&A]
Is there any reason why `lang.predicates.append(some_predicate_object)` shouldn't work? I'm confused as to why it doesn't...

**Comment by miquelramirez** [comment]
Hi @haz ,

back at the coal mine after a long weekend....

Could you please provide a snippet (over here or via gist) that illustrates the crash? exception? that you're getting?

Regards,

Miquel.

**Reply by haz** [reply]
That's just it... Nothing happens. I'm perplexed. The list of predicates is immutable somehow?

**Reply by miquelramirez** [reply]
Hi @haz,

okay, let me match your perplexity with my perplexity then when I understood what the problem you have is.

`lang.predicates` is a property, defined as follows in `src/tarski/fol.py`

```python
    @property
    def predicates(self):
        return list(self._predicates.values())
```

so of course adding a new predicate object to what clearly is a temporary object isn't going to pan out. The correct interface to define predicates is the method `predicate`

```
    def predicate(self, name: str, *args):
        self._check_name_not_defined(name, self._predicates, err.DuplicatePredicateDefinition)

        types = [self._retrieve_sort(a) for a in args]  # Convert possible strings into Sort objects
        predicate = Predicate(name, self, *types)
        self._predicates[name] = predicate
        self._global_index[name] = predicate
        return predicate
```

so you write 

```
foo = lang.predicate('foo', L.Object, L.Object)
```

to define a binary predicate where arguments of the predicate are to be constants of the built-in type `Object`. This is the way to define predicates in every test and documentation we have written like [this one](https://github.com/aig-upf/tarski/blob/master/docs/notebooks/defining-your-first-planning-problem.ipynb). As becomes apparent from the implementation of the `FirstOrderLanguage.predicate()` method, there's some significant bookkeeping that needs to be done in order to ensure that the definition of the new predicate is consistent with 1) that of other predicates and 2) the sorts defined to arrange the objects in the domain theory (or "first order language" as we chose to call it).

Miquel.

**Reply by haz** [reply]
Aha! Hadn't considered it being generated on the fly. Was so confused...

We worked around it as you suggested, but for the record, we were trying to merge domains that come with the exact same sorts (types and objects are unified, but predicates differ). Attempt was to just add the predicates of one domain, to the other, but we ended up creating them each from scratch.

Mystery solved, thanks!

**Reply by miquelramirez** [reply]
You need to copy the data from theory to theory. Theory merging is an interesting use case.

**Reply by haz** [reply]
It'll drop as an ICAPS demo and/or workshop paper. You're gonna like this one for sure ;)
